components/wifi/esp8266-driver/ESP8266/ESP8266.cpp

```diff
@@ -576,31 +577,50 @@ nsapi_error_t ESP8266::send(int id, const void *data, uint32_t amount)
     _smutex.lock();
     set_timeout(ESP8266_SEND_TIMEOUT);
     _busy = false;
-    if (_parser.send("AT+CIPSEND=%d,%lu", id, amount)
-            && _parser.recv(">")
-            && _parser.write((char *)data, (int)amount) >= 0
-            && _parser.recv("SEND OK")) {
-        // No flow control, data overrun is possible
-        if (_serial_rts == NC) {
-            while (_parser.process_oob()); // Drain USART receive register
-        }
-        _smutex.unlock();
-        return NSAPI_ERROR_OK;
+    _error = false;
+    if (!_parser.send("AT+CIPSEND=%d,%lu", id, amount)) {
+        tr_debug("ESP8266::send(): AT+CIPSEND failed");
+        goto END;
     }
-    if (_error) {
-        _error = false;
+
+    if(!_parser.recv(">")) {
+        tr_debug("ESP8266::send(): didn't get \">\"");
+        ret = NSAPI_ERROR_WOULD_BLOCK;
+        goto END;
     }
+
+    if (_parser.write((char *)data, (int)amount) >= 0 && _parser.recv("SEND OK")) {
+        ret = NSAPI_ERROR_OK;
+    }
+
+END:
```

Comment:
Instead of using the goto statement, the ifs above could be transformed into an if - else if - else if structure.

Reply:
Duly noted, but I'll keep the original for now.

Reply:
I kind of like the `goto END` type of handling where we have multiple cases which all should lead to shared error handling.
Its like `try { ... } catch { ... } finally { ... }`

Reply:
In a previous life, I worked with `do { ... } while (false)`. I've always had mixed feelings about it.